frame/offences/src/tests.rs

```diff
@@ -245,3 +281,157 @@ fn should_properly_count_offences() {
+			offenders: vec![6, 7],
+		};
+		let offence4 = Offence {
+			validator_set_count: 5,
+			session_index: session_index - 1,
+			time_slot: time_slot - 1,
+			offenders: vec![3],
+		};
+		Offences::report_offence(vec![], offence1).unwrap();
+		with_on_offence_fractions(|f| {
+			assert_eq!(f.clone(), vec![Perbill::from_percent(25)]);
+			f.clear();
+		});
+
+		// when
+		// report for the second time
+		Offences::report_offence(vec![], offence2).unwrap();
+		Offences::report_offence(vec![], offence3).unwrap();
+		Offences::report_offence(vec![], offence4).unwrap();
+
+		// then
+		let prev_session_reports = Vec::<(Kind, OpaqueTimeSlot, sp_core::H256)>::decode(
+			&mut &crate::SessionReports::<crate::mock::Runtime>::get(session_index - 1)[..],
+		)
+		.unwrap();
+		assert_eq!(
+			prev_session_reports,
+			vec![(KIND, (time_slot - 1).encode(), report_id(time_slot - 1, 3)),]
+		);
+
+		let session_reports = Vec::<(Kind, OpaqueTimeSlot, sp_core::H256)>::decode(
+			&mut &crate::SessionReports::<crate::mock::Runtime>::get(session_index)[..],
+		)
+		.unwrap();
+		assert_eq!(
+			session_reports,
+			vec![
+				(KIND, time_slot.encode(), report_id(time_slot, 5)),
+				(KIND, (time_slot + 1).encode(), report_id(time_slot + 1, 4)),
+			]
+		);
+
+		let next_session_reports = Vec::<(Kind, OpaqueTimeSlot, sp_core::H256)>::decode(
+			&mut &crate::SessionReports::<crate::mock::Runtime>::get(session_index + 1)[..],
+		)
+		.unwrap();
+		assert_eq!(
+			next_session_reports,
+			vec![
+				(KIND, (time_slot + 1).encode(), report_id(time_slot + 1, 6)),
+				(KIND, (time_slot + 1).encode(), report_id(time_slot + 1, 7)),
+			]
+		);
+	});
+}
+
+#[test]
+fn should_properly_clear_obsolete_offences() {
+	new_test_ext().execute_with(|| {
+		// given
+		let time_slot = 42;
+		let session_index = 10;
+		assert_eq!(offence_reports(KIND, time_slot), vec![]);
+
+		let offence1 =
+			Offence { validator_set_count: 5, session_index, time_slot, offenders: vec![5] };
+		let offence2 = Offence {
+			validator_set_count: 5,
+			session_index,
+			time_slot: time_slot + 1,
+			offenders: vec![4],
+		};
+		let offence3 = Offence {
+			validator_set_count: 5,
+			session_index: session_index + 1,
+			time_slot: time_slot + 5,
+			offenders: vec![6, 7],
+		};
+		let offence4 = Offence {
+			validator_set_count: 5,
+			session_index: session_index - 1,
+			time_slot: time_slot - 1,
+			offenders: vec![3],
+		};
+		Offences::report_offence(vec![], offence1).unwrap();
+		with_on_offence_fractions(|f| {
+			assert_eq!(f.clone(), vec![Perbill::from_percent(25)]);
+			f.clear();
+		});
+
+		// when
+		// report for the second time
+		Offences::report_offence(vec![], offence2).unwrap();
+		Offences::report_offence(vec![], offence3).unwrap();
+		Offences::report_offence(vec![], offence4).unwrap();
+
+		assert_eq!(
+			offence_reports(KIND, time_slot - 1),
+			vec![OffenceDetails { offender: 3, reporters: vec![] }]
+		);
+
+		Offences::clear_obsolete_reports(session_index + 5);
```

Comment:
somewhat more pedantic would be to call into `<Offences as SessionChangeListener>::xxx`